File: JLYoutubeActivity/JLYoutubeActivity.m
//
//  JLYoutubeActivity.m
//  onekm
//
//  Created by IM049 on 2015. 10. 21..
//  Copyright © 2015년 ces. All rights reserved.
//

#import "JLYoutubeActivity.h"

static NSString *YoutubeURLSchema = @"youtube://";
//static NSString *YoutubeBaseURL = @"http://www.youtube.com/v/";
static NSString *YoutubeBaseURL = @"youtube://watch?v=";

@interface JLYoutubeActivity ()

@property (nonatomic, strong) NSURL *URL;;
@end

@implementation JLYoutubeActivity

- (NSString *)activityType {
    return NSStringFromClass([self class]);
}

- (NSString *)activityTitle {
    return @"YouTube";
}

- (UIImage *)activityImage {
    if ([UIImage respondsToSelector:@selector(imageNamed:inBundle:compatibleWithTraitCollection:)]) {
        return [UIImage imageNamed:@"JLYoutubeActivity.bundle/youtube" inBundle:[NSBundle bundleForClass:self.class] compatibleWithTraitCollection:nil];
    }
    else {
        return [UIImage imageNamed:@"JLYoutubeActivity.bundle/youtube-7"];
    }
}

- (BOOL)canPerformWithActivityItems:(NSArray *)activityItems {
    if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:YoutubeURLSchema]] == YES) {
        for (id activityItem in activityItems) {
            return [self isYouTubeURL:activityItem];
        }
    }
    return NO;
}

- (void)prepareWithActivityItems:(NSArray *)activityItems {
    for (id activityItem in activityItems) {
        if ([self isYouTubeURL:activityItem] == YES) {
            NSString *videoIdentifier = [self videoIdentifierForUrl:activityItem];
            
            if (videoIdentifier != nil) {
                self.URL = [NSURL URLWithString:[YoutubeBaseURL stringByAppendingString:videoIdentifier]];
                break;
            }
        }
    }
}

- (void)performActivity {
    
    if ([[UIApplication sharedApplication] respondsToSelector:@selector(openURL:options:completionHandler:)]) {
        [[UIApplication sharedApplication] openURL:self.URL
                                           options:@{}
                                 completionHandler:^(BOOL success) {
                                     [self activityDidFinish:success];
                                 }];
    }
    else {
        BOOL completed = [[UIApplication sharedApplication] openURL:self.URL];
        [self activityDidFinish:completed];
    }
}

- (NSString*)videoIdentifierForUrl:(NSURL*)url {
    if ([url isKindOfClass:[NSURL class]]) {
        NSDictionary *parameters = [self queryParametersForURL:url];
        if (parameters[@"v"] != nil) {
            return parameters[@"v"];
        }
    }
    return nil;
}

- (BOOL)isYouTubeURL:(NSURL*)url {
    if ([url isKindOfClass:[NSURL class]]) {
        NSDictionary *parameters = [self queryParametersForURL:url];
        if ([parameters[@"feature"] isEqualToString:@"youtu.be"]) {
            return YES;
        }
    }
    return NO;
}

- (NSDictionary*)queryParametersForURL:(NSURL*)url {
    NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
    for (NSString *param in [[url query] componentsSeparatedByString:@"&"])
    {
        NSArray *parts = [param componentsSeparatedByString:@"="];
        if([parts count] < 2) continue;
        [params setObject:[parts objectAtIndex:1] forKey:[parts objectAtIndex:0]];
    }
    return params;
}

@end

File: JLYoutubeActivity/JLYoutubeActivity.swift
//
//  JLYoutubeActivity.swift
//  JLYoutubeActivityDemo
//
//  Created by Woody on 2018. 4. 24..
//  Copyright © 2018년 Woody. All rights reserved.
//

import UIKit

class JLYoutubeActivity: UIActivity {
    var YoutubeURLSchema: String = "youtube://"
    var YoutubeBaseURL: String = "youtube://watch?v="
    
    var url:URL?
    // override activity
    
    override var activityType: UIActivityType? {
        return  UIActivityType(String(describing: self.self))
    }
    
    override var activityTitle: String?{
        return "Youtube"
    }
    
    override var activityImage: UIImage? {
        return UIImage(named: "JLYoutubeActivity.bundle/youtube", in: Bundle(for: JLYoutubeActivity.self), compatibleWith: nil)
    }
    
    override func canPerform(withActivityItems activityItems: [Any]) -> Bool {
        guard let opneURL = URL(string: YoutubeURLSchema) else { return false }
        
        if UIApplication.shared.canOpenURL(opneURL) {
            return activityItems.contains { (item) -> Bool in
                guard let url = item as? URL else { return false }
                return url.isYoutube
            }
        }
        return false
    }
    
    override func prepare(withActivityItems activityItems: [Any]) {        
        for item in activityItems {
            if let url = item as? URL, let identifier = url.videoIdentifier {
                self.url = URL(string: YoutubeBaseURL+identifier)
            }
        }
    }
    
    override func perform() {
        guard let url = url else { return }
        
        if #available(iOS 10.0, *) {
            UIApplication.shared.open(url, options: [String:Any]()) { (success) in
                self.activityDidFinish(success)
            }
        }
        else {
            UIApplication.shared.openURL(url)
        }
    }
    
}

extension URL {    
    fileprivate var videoIdentifier: String? {
        guard isYoutube != false else { return nil}
        
        if let identifier = parameters?["v"] {
            return identifier
        }
        return nil
    }
    
    fileprivate var isYoutube: Bool {
        if let feature = parameters?["feature"], feature == "youtu.be" {
            return true
        }
        return false
    }
    
    fileprivate var parameters: [String:String]? {
        guard let components = NSURLComponents(url: self, resolvingAgainstBaseURL: false) else { return nil }
        guard let items = components.queryItems, items.isEmpty == false else { return nil }
        
        var parameters = [String:String]()
        for item in items  {
            if let value = item.value {
                parameters[item.name] = value
            }
        }
        return parameters;
    }
    
}
